src/Storages/MergeTree/MergeTreeData.cpp

```diff
@@ -906,46 +905,108 @@ void MergeTreeData::loadDataParts(bool skip_sanity_checks)
+                {
+                    /// Create and correctly initialize global WAL object
+                    /// There should be only one wal file, so this can be assigned in thread pool.
+                    write_ahead_log = std::make_shared<MergeTreeWriteAheadLog>(*this, disk_ptr, it->name());
+                    for (auto && part : write_ahead_log->restore(metadata_snapshot, getContext()))
+                        disk_wal_parts.push_back(std::move(part));
+                }
+                else if (settings->in_memory_parts_enable_wal)
+                {
+                    MergeTreeWriteAheadLog wal(*this, disk_ptr, it->name());
+                    for (auto && part : wal.restore(metadata_snapshot, getContext()))
+                        disk_wal_parts.push_back(std::move(part));
+                }
             }
-        }
+        });
+    }
+
+    pool.wait();
+
+    for (auto & [_, disk_wal_parts] : disk_wal_part_map)
+        parts_from_wal.insert(
+            parts_from_wal.end(), std::make_move_iterator(disk_wal_parts.begin()), std::make_move_iterator(disk_wal_parts.end()));
+
+    size_t num_parts = 0;
+    std::queue<std::vector<std::pair<String, DiskPtr>>> parts_queue;
+    for (auto & [_, disk_parts] : disk_part_map)
+    {
+        num_parts += disk_parts.size();
+        parts_queue.push(std::move(disk_parts));
     }
 
+
     auto part_lock = lockParts();
     data_parts_indexes.clear();
 
-    if (part_names_with_disks.empty() && parts_from_wal.empty())
+    if (num_parts == 0 && parts_from_wal.empty())
     {
         LOG_DEBUG(log, "There are no data parts");
         return;
     }
 
-    /// Parallel loading of data parts.
-    size_t num_threads = std::min(size_t(settings->max_part_loading_threads), part_names_with_disks.size());
+    pool.setMaxThreads(std::min(size_t(settings->max_part_loading_threads), num_parts));
+    size_t num_threads = pool.getMaxThreads();
+    size_t min_parts_per_thread = (num_parts - 1) / num_threads + 1;
+
+    /// Prepare data parts for parallel loading. Threads will focus on given disk first, then steal
+    /// others' tasks when finish current disk part loading process.
+    std::vector<std::vector<std::pair<String, DiskPtr>>> threads_parts(num_threads);
+    std::set<size_t> remaining_thread_parts;
+    for (size_t i = 0; i < num_threads && !parts_queue.empty(); ++i)
+    {
+        remaining_thread_parts.insert(i);
+        auto need_parts = min_parts_per_thread;
+        auto & thread_parts = threads_parts[i];
+
+        while (need_parts > 0 && !parts_queue.empty())
+        {
+            auto & current_parts = parts_queue.front();
+            if (current_parts.size() <= need_parts)
+            {
+                thread_parts.insert(thread_parts.end(), current_parts.begin(), current_parts.end());
+                parts_queue.pop();
+                need_parts -= current_parts.size();
```

Comment:
`parts_queue.pop()` removes `current_parts` memory, and makes use-after-free, right?

Reply:
Ah, right. Thanks! I've rewritten it in a more intuitive way.

Reply:
Yep, now looks better.